Add New documentation on Using Incrementals (#3343)
jenkins-infra/jenkins.io: content/doc/developer/plugin-development/incrementals.adoc

is an error due to which incrementals version was not created.
One of the most likely reasons is that latest changes from master have not been pulled into the branch.
You can click on 'Details' adjacent to it to investigate this error from the logs.

== Copy the Version ID

You should be redirected to a page like below:

image::/images/developer/plugin-development/incrementals/incrementals.png[]

Copy the version ID, which in this example would be `2.238-rc29961.5c3c5871cca6`.

Instead, you can also run the following in the upstream repository to obtain the incrementals version:
```
mvn -Dset.changelist validate
```

== Update the plugin POM

There are two changes you want to make to your plugin POM.

First, update the Jenkins version to the version ID copied in the previous step:

```
...
    <properties>
        <jenkins.version>2.238-rc29961.5c3c5871cca6</jenkins.version>
        ...
    </properties>
...
```

In case of upstream plugin, add its version in the `<version>` of your `<dependency>` inside the downstream plugin.

Now you can run your project like you normally do (say using `hpi:run`), and Jenkins will it automatically run it

> against the incrementals version of Jenkins core you specified.

Reviewer: ```suggestion
against the incrementals version of the upstream project you specified.
```

---
layout: developersection

> title: "Developing plugins and Jenkins core parallely"

Author: Thanks! Done :)

> title: "Developing plugins and Jenkins core in parallel"

Reviewer 2: Not sure how to capture the comment from @jglick in the title, but here is one possible idea:
```suggestion
title: "Developing components in parallel"
```

tags:
- jenkins
- incrementals
- core
- plugin

> description: "This article explains how to use Incrementals for developing plugins parallely with Jenkins core"

Author: Thanks! Done :)

---

A common problem one might stumble upon is developing an API in Jenkins core and simultaneously
building a reference implementation inside a plugin.
What to do, one might ask.

Worry not, Incrementals are here to the rescue!

With incrementals, we can create Jenkins core incremental versions which can be used along with the plugin to build the

> plugin against the very latest changes you have made to core.

Reviewer: This is only scenario. Equally, perhaps more, important in practice is developing an API in a base plugin (an “upstream” PR) and consuming it in a more specific plugin (a “downstream” PR).

Reviewer 2: The example might be a new addition to an API plugin and a simultaneous change in a consuming plugin that uses the change in the API plugin.  For example, the GitHub plugin might need an addition in the GitHub API plugin or the Git plugin might need an addition in the Git Client plugin.

Also, it allows keeping track and switching between incremental versions.
Not only that, it allows others to also not have to build Jenkins locally to try out the changes or collaborate with
you on your awesome project!

Sounds cool? Let's see how to go about it.

> == 1. Create a Pull Request in Jenkins core

Reviewer 2: Headings usually are not numbered.  We rely on the table of contents and linear reading rather than numbering inside the heading
```suggestion
== Create a Pull Request in Jenkins core
```

> First step is to create a PR on github link:https://github.com/jenkinsci/jenkins/pulls[here] by clicking the "New Pull

Reviewer 3: please don't use the term 'here' in links it's not as clear as it could be, see https://uxmovement.com/content/why-your-links-should-never-say-click-here/, it's also not very accessible content

A link should describe where it's going to be the text of the page, i.e. be more descriptive than just writing here

possibly this would be a good option:
```suggestion
First step is to create a PR on link:https://github.com/jenkinsci/jenkins/pulls[GitHub] by clicking the "New Pull
```

Author: Thanks Tim for the article, nice read :) I have made the requested changes

Request" button and push the changes to Jenkins core.

image::/images/developer/plugin-development/incrementals/create-pr.png[]

> Note: Make sure that you rebase your branch with the latest changes from master, otherwise incrementals version

Reviewer: You can simply `git merge` the origin `master` branch; it is necessary to rebase.

Reviewer 2: Agreed with @jglick that this `Note` can be deleted.  Any commit to the branch (either by merge or by new changes) will allow a new incremental version to be delivered.

will NOT be created.

> == 2. Wait for all checks to pass

Reviewer 2: Headings not usually numbered
```suggestion
== Wait for all checks to pass
```

> Wait for 'continuous-integration/jenkins' to perform checks, this can take about two hours.

Reviewer: Again, if the upstream PR is on a plugin, this would typically be much faster.

> == 3. Obtain the Incrementals version

Reviewer 2: ```suggestion
== Obtain the Incrementals version
```

After the checks are completed, you should get a message like below:

image::/images/developer/plugin-development/incrementals/checks.png[]

Click on 'Details' adjacent to 'continuous-integration/jenkins/incrementals — Deployed to Incrementals.'

> Note: If you only see one check 'continuous-integration/jenkins/pr-merge — This commit looks good' then there is an

Reviewer 2: ```suggestion
Note: If you only see one GitHub check 'continuous-integration/jenkins/pr-merge — This commit looks good' then there is an
```

error due to which incrementals version was not created. One of the most likely reasons is that latest changes from
master have not been pulled into the branch. You can click on 'Details' adjacent to it to investigate this error from
the logs.

> == 4. Copy the Version ID

Reviewer 2: ```suggestion
== Copy the Version ID
```

You should be redirected to a page like below:

> image::/images/developer/plugin-development/incrementals/4.png[]

Author: @timja Thanks! done :)

> Copy the version ID, which in this example would be `2.238-rc29961.5c3c5871cca6`.

Reviewer: Note you can also run in the upstream repository:

```bash
mvn -Dset.changelist validate
```

> == 5. Update the plugin POM

Reviewer 2: ```suggestion
== Update the plugin POM
```

There are two changes you want to make to your plugin POM.

First, update the Jenkins version to the version ID copied in the previous step:

```
...
    <properties>

>         <jenkins.version>2.238-rc29961.5c3c5871cca6</jenkins.version>

Reviewer: Again, this is only applicable to the case of Jenkins core being upstream. For a plugin upstream, you would just use that version in the `<version>` of your `<dependency>`.

        ...
    </properties>
...
```

> Then, add the incrementals repository to the list of repositories:

Reviewer 3: isn't this handled by the parent pom? / the incrementals plugin?

Author: Hi @timja! It wasn't handled automatically when I tried it, which eventually led me to write this guide :P

Reviewer 3: from what I understand it should be handled if the plugin is incrementalised  https://github.com/jenkinsci/incrementals-tools

but it doesn't work well for me locally in intellij (issues with dependency loading IIRC), I use SNAPSHOTs and then swap to incrementals for the CI build, but maybe this would work better

cc @jglick

Author: @timja I had a question actually (apologies if it is very simple) isn't this link for plugins so the plugins can create incrementals versions? In this particular guide I have not aimed at creating incrementals for the plugin, but for the Jenkins core version to use against.

Reviewer 3: isn't it the same steps? or does it not work right?

Author: Personally, it did not work for me, i got an error ` Unexpected tag: plugin-3.50`.

Reviewer: Do not do this! Having `-Pconsume-incrementals` in `.mvn/maven.config` (done for you by `mvn incrementals:incrementalify`) is all you need.

If IDEA does not honor this flag, that is a bug in your IDE, which you may be able to work around by manually configuring the IDEA project to use this profile. NetBeans handles this as of https://github.com/apache/netbeans/pull/1209 FWIW.

```
...
    <repositories>
        ...
        <repository>
            <id>incrementals.jenkins-ci.org</id>
            <url>https://repo.jenkins-ci.org/incrementals/</url>
        </repository>
        ...
    </repositories>
...
```

Now you can run your project like you normally do (say using `hpi:run`), and Jenkins will it automatically run it
against the incrementals version of Jenkins core you specified.

> == 6. (Optional) Enable usage of Beta API

Reviewer 2: ```suggestion
== (Optional) Enable usage of Beta API
```

> == 6. (Optional) Enable usage of Beta API
> 
> If the API you are building or are going to try has been marked
> link:https://www.jenkins.io/blog/2018/04/30/using-the-beta-annotation/[beta], then you would also have to add the
> beta tag, like so:
> 
> ```
> ...
>     <properties>
>         <jenkins.version>2.238-rc29961.5c3c5871cca6</jenkins.version>
>         <useBeta>true</useBeta>
>         ...
>     </properties>
> ...
> ```

Reviewer: I would drop this section. Not related to Incrementals at all, and not great advice to offer to the world—the idea is that only someone _in contact with_ the author of the beta API (perhaps the very same person) should be calling it, in which case you would be told to use this flag in a dedicated message.

```suggestion
```
